File: src/main/java/com/wesker926/leetcode/algorithms/p1163/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2023/4/24
// @description: 后缀数组
// go语言自带后缀数组SA-IS算法实现
package s1

import (
	"index/suffixarray"
	"unsafe"
)

func lastSubstring(s string) string {
	sa := (*struct {
		_  []byte
		sa []int32
	})(unsafe.Pointer(suffixarray.New([]byte(s)))).sa
	return s[sa[len(s)-1]:]
}

File: src/main/java/com/wesker926/leetcode/algorithms/p1163/s1/Solution.java
package com.wesker926.leetcode.algorithms.p1163.s1;

/**
 * @author wesker.gh
 * @date 2023/4/24
 * @description 双指针
 */
public class Solution {
    public String lastSubstring(String s) {
        int i = 0, n = s.length();
        char[] cs = s.toCharArray();
        for (int j = 1, k; j < n; ) {
            for (k = 0; j + k < n && cs[i + k] == cs[j + k]; k++) ;
            if (j + k < n && cs[i + k] < cs[j + k]) {
                int t = i;
                i = j;
                j = Math.max(j + 1, t + k + 1);
            } else {
                j = j + k + 1;
            }
        }
        return new String(cs, i, n - i);
    }
}
